R/utils.R

```diff
@@ -122,76 +122,124 @@ report_card_template <- function(title, label, description = NULL, with_filter,
-        html = function(with_module_name = TRUE) {
-          tagList(
-            lapply(
-              result,
-              function(x) x$html(with_module_name = with_module_name)
-            )
-          )
-        }
+  check_datanames <- check_modules_datanames_recursive(modules, datanames)
+  if (length(check_datanames)) {
+    modules_msg <- sapply(check_datanames, function(mod) {
+      sprintf(
+        "%s %s %s missing for module %s.",
+        if (length(mod$missing_datanames) > 1) "Datasets" else "Dataset",
+        toString(dQuote(mod$missing_datanames, q = FALSE)),
+        if (length(mod$missing_datanames) > 1) "are" else "is",
+        toString(dQuote(mod$label, q = FALSE))
       )
-    } else {
-      extra_datanames <- setdiff(modules$datanames, c("all", datanames))
-      if (length(extra_datanames)) {
-        list(
-          string = build_datanames_error_message(
-            modules$label,
-            datanames,
-            extra_datanames,
-            tags = list(
-              span = function(..., .noWS = NULL) { # nolint: object_name
-                trimws(paste(..., sep = ifelse(is.null(.noWS), " ", ""), collapse = " "))
-              },
-              code = function(x) toString(dQuote(x, q = FALSE))
+    })
+    sprintf(
+      "%s Datasets available in data: %s",
+      paste(modules_msg, collapse = "\n"),
+      toString(dQuote(datanames, q = FALSE))
+    )
+  } else {
+    TRUE
+  }
+}
+
+#' @rdname check_modules_datanames
+check_modules_datanames_html <- function(modules,
+                                         datanames) {
+  check_datanames <- check_modules_datanames_recursive(modules, datanames)
+  show_module_info <- inherits(modules, "teal_modules")
+  if (!length(check_datanames)) {
+    return(TRUE)
+  }
+  shiny::tagList(
+    lapply(
+      check_datanames,
+      function(mod) {
+        tagList(
+          tags$span(
+            tags$span(`if`(length(mod$missing_datanames) > 1, "Datasets", "Dataset")),
+            to_html_code_list(mod$missing_datanames),
+            tags$span(
+              paste0(
+                `if`(length(mod$missing_datanames) > 1, "are missing", "is missing"),
+                `if`(show_module_info, sprintf(" for tab '%s'.", mod$label), ".")
+              )
             ),
-            tagList = function(...) trimws(paste(...))
+            if (length(mod$datanames) >= 1) {
+              tagList(
+                tags$span(`if`(length(mod$datanames) > 1, "Datasets", "Dataset")),
+                tags$span("available in data:"),
+                tagList(
+                  tags$span(
+                    to_html_code_list(mod$datanames),
+                    tags$span(".", .noWS = "outside"),
+                    .noWS = c("outside")
+                  )
+                )
+              )
+            } else {
+              tags$span("No datasets are available in data.")
+            }
```

Comment:
The messages are different between text & HTML.

This corrects the problem and makes them equivalent.

```suggestion
            if (length(datanames) >= 1) {
              tagList(
                tags$span(if (length(datanames) > 1) "Datasets" else "Dataset"),
                tags$span("available in data:"),
                tagList(
                  tags$span(
                    to_html_code_list(datanames),
                    tags$span(".", .noWS = "outside"),
                    .noWS = c("outside")
                  )
                )
              )
            } else {
              tags$span("No datasets are available in data.")
            }
```